packages/pyright-internal/src/analyzer/semanticTokensWalker.ts

```diff
@@ -548,14 +555,69 @@ export class SemanticTokensWalker extends ParseTreeWalker {
                 modifiers.push(CustomSemanticTokenModifiers.classMember);
 
                 const declaredType = this._evaluator.getTypeForDeclaration(decl)?.type;
-                // the canonical check for properties (used e.g. in the hover message)
-                if (declaredType && isMaybeDescriptorInstance(declaredType)) {
-                    if (declarations.every((d) => this._missingPropertySetter(d))) {
-                        modifiers.push(SemanticTokenModifiers.readonly);
+
+                // if there is no type information or it is not (potentially) a descriptor instance,
+                // use the default `method` token type
+                if (!declaredType || !isMaybeDescriptorInstance(declaredType)) {
+                    return SemanticTokenTypes.method;
+                }
+
+                const isProp = isProperty(declaredType);
+                // if the descriptor instance is a property, check the declarations for `fset` information
+                // if it is not a property, use a check for the presence of `__set__`
+                const isReadOnly = isProp
+                    ? declarations.every((d) => this._missingPropertySetter(d))
+                    : !isMaybeDescriptorInstance(declaredType, true);
+                if (isReadOnly) {
+                    modifiers.push(SemanticTokenModifiers.readonly);
+                }
+
+                const isWrite = isWriteAccess(node);
+                let funType: Type | undefined;
+                if (isProp) {
+                    // since `__get__`/`__set__` cannot be found for properties, try to find the first
+                    // `fget`/`fset` information and use its method type if it exists
+                    funType = declarations
+                        .map((decl) => this._getPropertyInfo(decl, isWrite))
+                        .find((i) => i)?.methodType;
+                } else {
+                    // for non-properties, get the type of `__get__`/`__set__` of the descriptor instance
+                    const memberName = isWrite ? '__set__' : '__get__';
+                    const method = isWrite ? 'set' : 'get';
+                    const member = isClass(declaredType)
+                        ? this._evaluator.getTypeOfBoundMember(node, declaredType, memberName, { method: method })
+                        : undefined;
+                    funType = member?.type;
+                }
+                let effType: Type | undefined = undefined;
```

Comment:
nit: for readability i like to avoid abbreviating variable names. can we rename these variables to `functionType` and `effectiveType` instead